app/services/copy_course.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+#= Copy course from another server dashboard
+class CopyCourse
+  def initialize(url:)
+    @url = url
+  end
+
+  def make_copy
+    @course_data = retrieve_course_data
+    copy_main_course_data
+    add_tracked_wikis
+    @cat_data = retrieve_categories_data
+    copy_tracked_categories_data
+    @users_data = retrieve_users_data
+    copy_users_data
+    return { course: @course, error: nil }
+  rescue ActiveRecord::RecordInvalid, StandardError => e
+    return { course: nil, error: e.message }
+  end
+
+  private
+
+  def copy_main_course_data
+    # Extract the attributes we want to copy
+    params_to_copy = %w[school title term description start end subject slug timeline_start
+                        timeline_end type flags]
+    copied_data = {}
+    params_to_copy.each { |p| copied_data[p] = @course_data[p] }
+    @home_wiki = Wiki.get_or_create(language: @course_data['home_wiki']['language'],
+                                    project: @course_data['home_wiki']['project'])
+    copied_data['home_wiki_id'] = @home_wiki.id
+    copied_data['passcode'] = 'passcode' # set an arbitrary passcode
```

Comment:
Since we could potential use this in production to copy from one server to another, I think it will be better to use a random passcode via `GeneratePasscode.call`.

Reply:
Done on 65188dad85135cd0f29871dc2195194f4d13eb81